code/game/objects/items/devices/gps.dm

```diff
@@ -1,28 +1,423 @@
+			continue
+	
+		var/turf/gps_turf = get_turf(gps)
+		var/gps_tag = LAZYACCESS(showing_tracked_names, thing) ? gps.gps_tag : null
+		if(istype(gps_turf))
+			compass.set_waypoint("\ref[gps]", gps_tag, gps_turf.x, gps_turf.y, gps_turf.z, LAZYACCESS(tracking_devices, "\ref[gps]"))
+			if(can_track(gps) && my_turf && gps_turf != my_turf)
+				compass.show_waypoint("\ref[gps]")
+	
+	compass.rebuild_overlay_lists(update_compass_icon)
+
+/obj/item/gps/AltClick(mob/user)
+	toggletracking(user)
+
+/obj/item/gps/proc/toggletracking(mob/living/user)
+	if(!istype(user))
+		return
+
+	if(emped)
+		to_chat(user, SPAN_WARNING("\The [src] is busted!"))
+		return
+	
+	toggle_tracking()
+	if(tracking)
+		to_chat(user, SPAN_NOTICE("\The [src] is now tracking, and visible to other GPS devices."))
+	else
+		to_chat(user, SPAN_NOTICE("\The [src] is no longer tracking, or visible to other GPS devices."))
+
+/obj/item/gps/proc/toggle_tracking()
+	tracking = !tracking
+	if(tracking)
+		if(!is_in_processing_list)
+			is_in_processing_list = TRUE
+			START_PROCESSING(SSobj, src)
+	else
+		is_in_processing_list = FALSE
+		STOP_PROCESSING(SSobj, src)
+	update_compass()
+	update_holder()
+	update_icon()
+
+/obj/item/gps/emp_act(severity)
+	if(emped) // Without a fancy callback system, this will have to do.
+		return
+	var/severity_modifier = severity ? severity : 4 // In case emp_act gets called without any arguments.
+	var/duration = 5 MINUTES / severity_modifier
+	emped = TRUE
+	update_icon()
+	addtimer(CALLBACK(src, .proc/reset_emp), duration)
+
+/obj/item/gps/proc/reset_emp()
+	emped = FALSE
+	update_icon()
+	if(ismob(loc))
+		to_chat(loc, SPAN_NOTICE("\The [src] appears to be functional again."))
+
+/obj/item/gps/on_update_icon()
+
+	cut_overlays()
+
+	var/image/I
+	for(var/decal in decals)
+		if(check_state_in_icon(decal, decal_icon))
+			I = image(decal_icon, decal)
+			I.color = decals[decal]
+			I.appearance_flags |= RESET_COLOR
+
+	if(screen_color && check_state_in_icon("[icon_state]-keypad", decal_icon))
+		I = image(decal_icon, "[icon_state]-keypad")
+		I.color = screen_color
+		I.appearance_flags |= RESET_COLOR
+		add_overlay(I)
+
+	if(emped)
+		if(check_state_in_icon("[icon_state]-emp", decal_icon))
+			I = image(decal_icon, "[icon_state]-emp")
+			I.color = screen_color
+			I.appearance_flags |= RESET_COLOR
+			add_overlay(I)
+	else if(tracking)
+		if(check_state_in_icon("[icon_state]-working", decal_icon))
+			I = image(decal_icon, "[icon_state]-working")
+			I.color = screen_color
+			I.appearance_flags |= RESET_COLOR
+			add_overlay(I)
+	else if(check_state_in_icon("[icon_state]-screen", decal_icon))
+		I = image(decal_icon, "[icon_state]-screen")
+		I.color = screen_color
+		I.appearance_flags |= RESET_COLOR
+		add_overlay(I)
+
+/obj/item/gps/attack_self(mob/user)
+	display(user)
+
+ // Compiles all the data not available directly from the GPS
+ // Like the positions and directions to all other GPS units
+/obj/item/gps/proc/display_list()
```

Comment:
Yeah, not really getting the UI model here. Separating this from the template layer maybe makes sense when using nanoUI, but why here? You're bypassing a bunch of safety stuff, potentially.

Reply:
It's honestly nonsensical. I think I got sick of the revision on Polaris and they didn't review it thoroughly at all.